### PR TITLE
Exclude more code from custom_ops.cc when not required in minimal build

### PR DESCRIPTION
**Pull request by skottmckay**
### Description

- Split out the code that implements the OrtKernelContext API (used by compiled nodes and custom ops) and the code that implements the custom ops API.
- Exclude based on minimal build settings using helpers
  - the main change is to simply wrap the implementation into a lambda so it can be easily enabled/disabled
  - actual implementation of all functions are unchanged
- Re-organize so the related implementations are together
  - most diffs are from this, but without the reorg it would be much harder to know which helper to use
- General cleanup of lines that were too long.

### Motivation and Context

Saves ~10KB in a minimal build.

Build command used for comparison
```
./build --android --android_api=29 --android_sdk="d:\Android" --android_abi=arm64-v8a --parallel --android_ndk_path="D:\Android\ndk\26.0.10792818\" --build_shared_lib --cmake_generator Ninja --skip_tests --minimal_build --disable_rtti --disable_ml_ops --disable_exceptions --cmake_extra_defines=onnxruntime_BUILD_UNIT_TESTS=OFF --include_ops_by_config .\no_ops.config --config MinSizeRel
```

Main: 1,218,480 bytes
With changes: 1,208,320 bytes

### CODE REVIEW

File: onnxruntime/core/session/custom_ops.cc [hunk 1]
```diff
@@ -121,304 +131,392 @@ struct OrtShapeInferContext {
 };
 #endif
 
-ORT_API_STATUS_IMPL(OrtApis::ShapeInferContext_GetInputCount, _In_ const OrtShapeInferContext* context, _Out_ size_t* out) {
+#if ENABLE_ORT_KERNEL_CONTEXT_API
+template <typename T>
+static OrtStatusPtr ExecuteIfKernelContextApiEnabled(const T& fn) {
```
Comment (skottmckay) [inline onnxruntime/core/session/custom_ops.cc:136]: I tried using `const std::function<OrtStatusPtr(void)>& fn` initially but that resulted in binary size growth. Using a template seems to allow the compiler/linker to successfully throw away the unused implementation. This occurred with both Windows and Android builds.

File: onnxruntime/core/session/custom_ops.cc [hunk 2]
```diff
@@ -121,304 +131,392 @@ struct OrtShapeInferContext {
...
-  API_IMPL_END
-};
+    switch (type) {
+      case OrtOpAttrType::ORT_OP_ATTR_FLOAT: {
+        if (len < sizeof(float)) {
+          ret = OrtApis::CreateStatus(OrtErrorCode::ORT_INVALID_ARGUMENT,
+                                      "Size of data not large enough to hold a float.");
+        } else {
+          if (attr->has_f()) {
+            auto output_f = reinterpret_cast<float*>(data);
```
Comment (edgchen1) [inline onnxruntime/core/session/custom_ops.cc:376]: not related to this PR, but do we need to consider whether `data` is correctly aligned before writing something like a float to it?

**Review by edgchen1 — APPROVED** [review]

File: onnxruntime/core/session/custom_ops.cc (hunk 1 above)
Comment (skottmckay) [reply onnxruntime/core/session/custom_ops.cc:136]: Not sure. Will leave for now but consider removing if I touch this file again.

File: onnxruntime/core/session/custom_ops.cc [hunk 3]
```diff
@@ -121,304 +131,392 @@ struct OrtShapeInferContext {
...
-    if (len < sizeof(int)) {
-      ret = OrtApis::CreateStatus(OrtErrorCode::ORT_INVALID_ARGUMENT, "Size of data not large enough to hold an int64.");
-    } else {
-      if (attr->has_i()) {
-        auto output_i = reinterpret_cast<int64_t*>(data);
-        *output_i = attr->i();
+ORT_API_STATUS_IMPL(OrtApis::KernelContext_ParallelFor, _In_ const OrtKernelContext* context,
+                    _In_ void (*fn)(void*, size_t), _In_ size_t total, _In_ size_t num_batch, _In_ void* usr_data) {
+  return ExecuteIfKernelContextApiEnabled([&]() -> OrtStatusPtr {
+    if (!context) {
```
[the remark this answers is not in the document]
Comment (skottmckay) [reply onnxruntime/core/session/custom_ops.cc:225]: Not sure. Will leave for now but consider removing if I touch this file again.

**Pull request merged**